samza-core/src/main/java/org/apache/samza/container/host/OshiBasedStatisticsGetter.java

```diff
@@ -0,0 +1,90 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.container.host;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.NotImplementedException;
+import oshi.SystemInfo;
+import oshi.software.os.OSProcess;
+import oshi.software.os.OperatingSystem;
+
+
+/**
+ * An implementation of {@link SystemStatisticsGetter} that relies on using oshi framework(https://www.oshi.ooo/)
+ */
+public class OshiBasedStatisticsGetter implements SystemStatisticsGetter {
+  // the snapshots of current JVM process and its child processes
+  private final Map<Integer, OSProcess> previousProcessSnapshots = new HashMap<>();
+
+  private final OperatingSystem os;
+  private final int cpuCount;
+
+  public OshiBasedStatisticsGetter() {
+    this(new SystemInfo());
+  }
+
+  @VisibleForTesting
+  OshiBasedStatisticsGetter(SystemInfo si) {
+    this(si.getOperatingSystem(), si.getHardware().getProcessor().getPhysicalProcessorCount());
+  }
+
+  @VisibleForTesting
+  OshiBasedStatisticsGetter(OperatingSystem os, int cpuCount) {
+    this.os = os;
+    this.cpuCount = cpuCount;
+  }
+
+  @Override
+  public SystemMemoryStatistics getSystemMemoryStatistics() {
+    throw new NotImplementedException("Not implemented");
+  }
+
+  @Override
+  public ProcessCPUStatistics getProcessCPUStatistics() {
+    final List<OSProcess> currentProcessAndChildProcesses = getCurrentProcessAndChildProcesses();
+    final double totalCPUUsage = getTotalCPUUsage(currentProcessAndChildProcesses);
+    refreshProcessSnapshots(currentProcessAndChildProcesses);
+    return new ProcessCPUStatistics(100d * totalCPUUsage / cpuCount);
+  }
+
+  private List<OSProcess> getCurrentProcessAndChildProcesses() {
+    final List<OSProcess> processes = new ArrayList<>();
+    // get current process
+    processes.add(os.getProcess(os.getProcessId()));
+    // get all child processes of current process
+    processes.addAll(os.getChildProcesses(os.getProcessId(), OperatingSystem.ProcessFiltering.ALL_PROCESSES,
+        OperatingSystem.ProcessSorting.NO_SORTING, 0));
+    return processes;
+  }
+
+  private double getTotalCPUUsage(List<OSProcess> processes) {
+    return processes.stream()
+        .mapToDouble(p -> p.getProcessCpuLoadBetweenTicks(previousProcessSnapshots.get(p.getProcessID())))
```

Comment:
No, it won't: https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/software/common/AbstractOSProcess.java#L52

```java
    @Override
    public double getProcessCpuLoadBetweenTicks(OSProcess priorSnapshot) {
        if (priorSnapshot != null && this.processID == priorSnapshot.getProcessID()
                && getUpTime() > priorSnapshot.getUpTime()) {
            return (getUserTime() - priorSnapshot.getUserTime() + getKernelTime() - priorSnapshot.getKernelTime())
                    / (double) (getUpTime() - priorSnapshot.getUpTime());
        }
        return getProcessCpuLoadCumulative();
    }

```

Reply:
actually this brings out an important point:
the impl is 
```
public ProcessCPUStatistics getProcessCPUStatistics() {
    final List<OSProcess> currentProcessAndChildProcesses = getCurrentProcessAndChildProcesses();
    final double totalCPUUsage = getTotalCPUUsage(currentProcessAndChildProcesses);
    refreshProcessSnapshots(currentProcessAndChildProcesses);
    return new ProcessCPUStatistics(100d * totalCPUUsage / cpuCount);
```
so the first time this is called, `previousProcessSnapshots` is an empty hash map right. so `get` from it will return null. seems in this case the cpu usage returned cumulative in that case. which seems okay.

Reply:
Yes, it will return cumulative data points. And I think it is reasonable and accurate because that's the first data point available in the process start-up phase.